### PR TITLE
Fixes duplication of reports and other events

### PR DESCRIPTION
**Pull request by karthiksbhat**
Events like report generation, bot creation etc were being duplicated because of two listeners set up to listen to the same events. There is still possibility of discovering a bug related to this in the future: there is a listener for all events emitted by the API process, but only for 1 event emitted by the Fetching process. But currently, this code fixes the known bugs related to this code. The changes to bot-master.js have been made to mimic Andres' PR (PR 171).
End to end tests: 
Some changes have been copied from Philip's other branch to avoid future merge conflicts.
Toolkit has some new useful functions which can be reused in other tests.
duplication-test.js: This is a reusable test which will add new source, listen for reports, and check if the correct number of reports are generated. This was designed to check duplication of reports.

### CODE REVIEW

File: test/end-to-end/duplication.test.js [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+      utils.addSource('SMS GH', { nickname: 'hello', keywords: 'test' });
+    })
+    .then(function() {
+      utils.toggleFetching('On');
+    })
+    .then(function() {
+      utils.toggleSource('SMS GH', 'On');
+    })
+    .then(function() {
+      request('http://localhost:1111')
```
Comment (andresfib) [inline test/end-to-end/duplication.test.js]: make a function of the SMS Gh request. You can also abstract the 5 test, so you only pass the status of the toggles and the expected reports to a function, and don't duplicate so much code

File: lib/fetching/bot-master.js [hunk 2]
```diff
@@ -107,15 +109,29 @@ BotMaster.prototype._addSourceListeners = function(emitter) {
 
   // Listen to `enable` event from the API process
   emitter.on('source:enable', function(source) {
+    // This should ensure fetching button does its job
```
Comment (andresfib) [inline lib/fetching/bot-master.js]: To a different PR

File: lib/fetching/bot-master.js [hunk 3]
```diff
@@ -107,15 +109,29 @@ BotMaster.prototype._addSourceListeners = function(emitter) {
...
+    if (!self.enabled) {
+      return;
+    }
     var bot = self._getBot(source._id);
     if (self.enabled && bot) bot.start();
   });
 
   // Listen to `disable` event from the API process
   emitter.on('source:disable', function(source) {
+    // This should ensure fetching button does its job
```
Comment (andresfib) [inline lib/fetching/bot-master.js]: To a different PR

**Comment by karthiksbhat** [comment]
Closing because the issues handled by this PR have been split and submitted as separate PR's, as suggested.
